lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,245 @@
+from flask import jsonify
+from http import HTTPStatus
+from logging import Logger
+
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.auth import Auth
+from pbench.server.api.resources import (
+    JSON,
+    Schema,
+    Parameter,
+    ParamType,
+    PostprocessError,
+)
+from pbench.server.api.resources.query_apis import CONTEXT, ElasticBase
+from pbench.server.database.models.datasets import Dataset, Metadata, MetadataError
+from pbench.server.database.models.users import User
+from pbench.server.database.models.template import Template, TemplateNotFound
+
+
+class IterationSamples(ElasticBase):
+    """
+    Create iteration samples aggregation based on a given run id.
+    """
+
+    WHITELIST_AGGS_FIELDS = ["run", "sample", "iteration", "benchmark"]
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(Parameter("run_id", ParamType.STRING, required=True),),
+        )
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        request is authorized for this dataset.
+
+        If the user has authorization to read the dataset, return the Dataset
+        object as CONTEXT so that the postprocess operation can proceed to get all
+        the iteration aggregations.
+        """
+        run_id = client_json.get("run_id")
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error(f"Dataset with Run ID {run_id!r} not found")
+            abort(HTTPStatus.NOT_FOUND, message="Dataset not found")
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                f"Dataset owner ID { dataset.owner_id!r} cannot be found in Users"
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For Iteration samples, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset.access)
+
+        # The dataset exists, and authenticated user has enough access so continue
+        # the operation with the appropriate CONTEXT.
+        return {"dataset": dataset, "run_id": run_id}
+
+
+class IterationSampleNamespaces(IterationSamples):
+    """
+    Iteration samples API that returns only unique namespaces for each
```

Comment:
I'm not really sure where the "unique" comes from here. We're returning all of the aggregatable (keyword) *fields* in the whitelisted subdocuments, each with a list of all actual values. I think the "unique" is at best redundant, and may imply some additional postprocessing that would not be at all obvious.

Reply:
As you said we are returning aggregatable fields in the whitelisted subdocuments, each with a list of all actual values, but these actual values are the bucket keys from the ES query result and so they are ought to be unique. So yeah the unique word could be just redundant but I mentioned to in the docstring if someone doesnt know that these values represent (bucket keys from ES)

Reply:
The word "unique" is used so often and so casually as to have become meaningless, and it is simply not helpful in this context.